engine/src/main/java/org/terasology/rendering/dag/RenderingGraph.java

```diff
@@ -0,0 +1,140 @@
+
+    public RenderingGraph(Context context) {
+        this.context = context;
+        this.nodeMap = Maps.newLinkedHashMap(); // insert order is important for now
+        this.config = context.get(Config.class);
+        this.stateManager = StateManager.createDefault(context);
+        this.analyzed = false;
+    }
+
+    public static RenderingGraph createDefault(Context context) {
+        RenderingGraph pipeline = new RenderingGraph(context);
+        pipeline.add("shadowMap", ShadowMapNode.class);
+        pipeline.add("worldReflective", WorldReflectionNode.class);
+        pipeline.add("backdrop", BackdropNode.class);
+        pipeline.add("skyBands", SkyBandsNode.class);
+        pipeline.add("objectsOpaque", ObjectsOpaqueNode.class);
+        pipeline.add("chunksOpaque", ChunksOpaqueNode.class);
+        pipeline.add("chunksAlphaReject", ChunksAlphaRejectNode.class);
+        pipeline.add("overlays", OverlaysNode.class);
+        pipeline.add("firstPersonView", FirstPersonViewNode.class);
+        pipeline.add("lightGeometry", LightGeometryNode.class);
+        pipeline.add("directionalLights", DirectionalLightsNode.class);
+        pipeline.add("chunksRefractiveReflective", ChunksRefractiveReflectiveNode.class);
+        pipeline.add("outline", OutlineNode.class);
+        pipeline.add("ambientOcclusionPasses", AmbientOcclusionPassesNode.class);
+        pipeline.add("prePostComposite", PrePostCompositeNode.class);
+        pipeline.add("simpleBlendMaterials", SimpleBlendMaterialsNode.class);
+        pipeline.add("lightShafts", LightShaftsNode.class);
+        pipeline.add("initialPostprocessing", InitialPostProcessingNode.class);
+        pipeline.add("downSampleSceneAndUpdateExposure", DownSampleSceneAndUpdateExposureNode.class); // TODO: a rename needed :)
+        pipeline.add("toneMapping", ToneMappedSceneNode.class);
+        pipeline.add("bloomPasses", BloomPassesNode.class);
+        pipeline.add("blurPasses", BlurPassesNode.class);
+        pipeline.add("finalPostProcessing", FinalPostProcessingNode.class);
+        pipeline.analyze();
+        return pipeline;
+    }
+
+    public void analyze() {
+        topologicalOrder = getTopologicalOrder();
+        stateManager.findStateChanges(topologicalOrder);
+        analyzed = true;
+    }
+
+    public void add(AbstractNode node) {
+        nodeMap.put(node.getIdentifier(), node);
+    }
+
+    public <T extends AbstractNode> void add(String identifier, Class<T> type) {
+        // Attempt constructor-based injection first
+        context.put(String.class, identifier);
+        T node = InjectionHelper.createWithConstructorInjection(type, context);
+        // Then fill @In fields
+        InjectionHelper.inject(node, context);
+        node.initialise();
+        this.add(type.cast(node));
+    }
+
+    public Node get(String identifier) {
+        return nodeMap.get(identifier);
+    }
+
+    public void process() throws NotAnalyzedException {
```

Comment:
No, what I'm suggesting is that the RenderGraph class only returns a list of nodes through the getNodesInTopologicalOrder() method. The RenderPipelineGenerator takes that list and generates a new list that also includes StateChange instances interleaved with Nodes. The RenderPipelineOptimizer takes that new list and generates the list of commands to execute during rendering.